iceoryx_posh/include/iceoryx_posh/popo/active_call_set.hpp

```diff
@@ -0,0 +1,196 @@
+    template <typename T, typename EventType, typename = std::enable_if_t<std::is_enum<EventType>::value>>
+    void detachEvent(T& eventOrigin, const EventType eventType) noexcept;
+
+    /// @brief Detaches an event. Hereby, the event is defined as a class T, the eventOrigin.
+    /// @note This method can be called from any thread concurrently without any restrictions!
+    /// @tparam[in] T type of the class which will signal the event
+    /// @param[in] eventType enum required to specify the type of event inside of eventOrigin
+    template <typename T>
+    void detachEvent(T& eventOrigin) noexcept;
+
+    /// @brief Returns the capacity of the ActiveCallSet
+    /// @return capacity of the ActiveCallSet
+    static constexpr uint64_t capacity() noexcept;
+
+    /// @brief Returns the size of the ActiveCallSet
+    /// @return size of the ActiveCallSet
+    uint64_t size() const noexcept;
+
+  protected:
+    ActiveCallSet(EventVariableData* eventVariable) noexcept;
+
+  private:
+    class Event_t;
+
+    void threadLoop() noexcept;
+    cxx::expected<uint64_t, ActiveCallSetError>
+    addEvent(void* const origin,
+             const uint64_t eventType,
+             const uint64_t eventTypeHash,
+             CallbackRef_t<void> callback,
+             TranslationCallbackRef_t translationCallback,
+             const cxx::MethodCallback<void, uint64_t> invalidationCallback) noexcept;
+    void removeEvent(void* const origin, const uint64_t eventType, const uint64_t eventTypeHash) noexcept;
+
+    void removeTrigger(const uint64_t index) noexcept;
+
+  private:
+    enum class NoEnumUsed
+    {
+        PLACEHOLDER = 0
+    };
+
+    class Event_t
+    {
+      public:
+        ~Event_t();
+
+        bool isEqualTo(const void* const origin, const uint64_t eventType, const uint64_t eventTypeHash) const noexcept;
+        bool resetIfEqualTo(const void* const origin, const uint64_t eventType, const uint64_t eventTypeHash) noexcept;
+        bool reset() noexcept;
+        void init(const uint64_t eventId,
+                  void* const origin,
+                  const uint64_t eventType,
+                  const uint64_t eventTypeHash,
+                  CallbackRef_t<void> callback,
+                  TranslationCallbackRef_t translationCallback,
+                  const cxx::MethodCallback<void, uint64_t> invalidationCallback) noexcept;
+        void executeCallback() noexcept;
+        bool isInitialized() const noexcept;
+
+      private:
+        static constexpr uint64_t INVALID_ID = std::numeric_limits<uint64_t>::max();
+
+        void* m_origin = nullptr;
+        uint64_t m_eventType = INVALID_ID;
+        uint64_t m_eventTypeHash = INVALID_ID;
+
+        CallbackPtr_t<void> m_callback = nullptr;
+        TranslationCallbackPtr_t m_translationCallback = nullptr;
+
+        uint64_t m_eventId = INVALID_ID;
+        cxx::MethodCallback<void, uint64_t> m_invalidationCallback;
```

Comment:
Since you iterate over an array of the type containing this variables, please consider if smaller types would also be sufficient in order to have as much data as possible on a cache line. It borders on premature optimization, though and a struct of arrays might be even better than the array of structs. Maybe we need an issue where we can collect all these places to benchmark for potential speed gains.

Reply:
@elBoberido I never iterate over this array! I iterate over an array of bools which then provide me with a list of id's and then I access those elements explicitly.

Reply:
I didn't meant the hot path, but when adding/removing events to the `ActiveWaitSet`. But contrary to the notification vector this path should not be hot.

Reply:
I refined it even more and the array is only iterated when adding and event. But this can also be removed if we would allow that reattaching an event would override the current callback of the event. I would be fine with this! But maybe the user is then surprised why the event uses a different callback but it would be an error in the user code.

Maybe as a compromise we could allow attaching events which are already attached but then print a warning stating that the best practice is to detach events before reattaching them and that the callback is overridden. Would you be fine with this @elBoberido @mossmaurice

Reply:
In the upcoming refactoring I will also adjust the `addEvent` method so that we acquire the unique event id from the eventorigin if the event was already attached (or something similar) and then this array is never iterated. (will add this as a todo)